pike/test/copychunk.py

```diff
@@ -826,3 +889,228 @@ def generic_mchan_ssc_test_case(self):
+        filepair = []
+        blocks = []
+        for i in range(0, num_sess):
+            src_filename = "src_multiple_session" + str(i) + ".txt"
+            dst_filename = "dst_multiple_session_same_dest.txt"
+            filepair.append((src_filename, dst_filename,
+                             0 + i * SIMPLE_5_CHUNKS_LEN))
+            block = _gen_random_test_buffer(SIMPLE_5_CHUNKS_LEN)
+            self._create_and_write(src_filename, block)
+            blocks.append(block)
+        self.logger.info("ssc same dest multiple session test, session number is %d iteration number is %d" % (
+            num_sess, num_iter))
+        self.generic_multiple_ssc_test_case(
+            filepair, SIMPLE_5_CHUNKS, blocks, num_iter=num_iter)
+
+    def test_multiple_ssc_file_with_writethrough_flag(self):
+        """
+        multiple server side copy operation which write_through
+        flag set when creating the destination file handle
+        """
+        num_sess = 3
+        num_iter = 1
+        # file pair element {src_filename, dst_filename, dst_offset}
+        filepair = []
+        blocks = []
+        for i in range(0, num_sess):
+            src_filename = "src_multiple_session" + str(i) + ".txt"
+            dst_filename = "dst_multiple_session" + str(i) + ".txt"
+            filepair.append((src_filename, dst_filename, 0))
+            block = _gen_random_test_buffer(SIMPLE_5_CHUNKS_LEN)
+            self._create_and_write(src_filename, block)
+            blocks.append(block)
+        self.logger.info("ssc test with write_through set, session number is %d iteration number is %d" % (
+            num_sess, num_iter))
+        self.generic_multiple_ssc_test_case(
+            filepair, SIMPLE_5_CHUNKS, blocks, write_throu=True, num_iter=num_iter)
+
+    def generic_ssc_boundary_test_case(self, block, chunks,
+                                       exp_error=None, ssc_res=[0, 0, 0]):
+        """
+        check situation that one of the chunks which contain numbers cross the value
+        of end of file or maximum values system can support
+        """
+        src_filename = "src_copy_chunk_cross_offset.txt"
+        dst_filename = "dst_copy_chunk_cross_offset.txt"
+        self._create_and_write(src_filename, block)
+
+        fh_src, fh_dst = self._open_src_dst(src_filename, dst_filename)
+        resume_key = self.chan.resume_key(fh_src)[0][0].resume_key
+        future = self._submit_ssc_async_request(
+            self.chan, fh_dst, resume_key, chunks)
+
+        if exp_error is None:
+            exp_error = pike.ntstatus.STATUS_SUCCESS
+        try:
+            with self.assert_error(exp_error):
+                result = future[0].result()
+        finally:
+            self.chan.close(fh_src)
+            self.chan.close(fh_dst)
+
+        self.assertIsInstance(
+            future[0].response.response.children[0].ioctl_output, pike.smb2.CopyChunkCopyResponse)
```

Comment:
maybe to simplify this section of code you could get a reference to the ioctl response

```
ioctl_resp = future[0].response.response.children[0]
self.assertIsInstance(ioctl_resp.ioctl_output, pike.smb2.CopyChunkCopyResponse)
self.assertEqual(ioctl_resp.chunks_written, ssc_res[0])
self.assertEqual(ioctl_resp.chunk_bytes_written, ssc_res[1])
self.assertEqual(ioctl_resp.total_bytes_written, ssc_res[2])
```